### PR TITLE
Simple Pinned favorites.

### PR DESCRIPTION
**Pull request by dbolack-ab**
## Description

Implements a version of Favorites/Bookmarks ( using the term Pin ) as discussed today. 

Changes:
 * adds a pinned array to the Mongo document, containing a list of usernames who have pinned it.
 * adds a UI element to share pages on the left of the share title ( a pushpin ) to toggle pinning the document. White is unpinned, red is pinned.
 * Will not allow an author of a Brew to pin it.
 * Adds pinned themes to the theme selection pulldown.

I am unsure that this is absolutely safe for the document, should a user be editing the document when someone chooses to pin it. Perhaps some stronger filtering?

### Reviewer Checklist

*Reviewers, refer to this list when testing features, or suggest new items *
- [ ] Verify new features are functional
  - [ ] Correctly pins a brew via /share
  - [ ] Reload shows brew still pinned
  - [ ] Correctly unpins a brew via /share
  - [ ] Reload shows brew no longer pinned
  - [ ] Pinned theme shows on the theme pulldown
- [ ] Verify old features have not broken
  - [ ] No race conditions with simultaneous editing
- [ ] Test for edge cases / try to break things
  - [ ] Ensure Author updates do not break pins.
- [ ] Identify opportunities for simplification and refactoring
- [ ] Check for code legibility and appropriate comments
- [ ] Ensure Tests are adequate

### CODE REVIEW

**Comment by ericscheid** [comment]
Please create an issue for this PR so we've got traceability. Saying "as discussed today" is insufficient

Also:  
The `pinned` state is stored in the brew itself? The brew that is owned by someone else? That you don't have co-author rights to?

Does updating the brew with pinned also affect the modifiedAt and version metadata?

Does cloning the brew clear the pinning user ids?

File: server/homebrew.model.js [hunk 1]
```diff
@@ -65,6 +67,25 @@ HomebrewSchema.statics.getByUser = async function(username, allowAccess=false, f
 	return brews;
 };
 
+HomebrewSchema.statics.setPin  = async function(username, shareId, pin) {
+	// Test for User Ownership of brew
+	const brew = await Homebrew.findOne({ shareId: shareId }, null).exec()
+		.catch((error)=>{});
+	if(brew) {
+		if(!brew.authors.includes(username)) {
```
Comment (calculuschild) [inline server/homebrew.model.js:102]: Your current implementation does not allow an author to pin their own brews, which is fine for this first implementation. However, the check for authorship happens at two places. Would it be OK to remove the check here, and leave it in the client GUI only? Then this function can focus purely on setting the property as requested, and we don't have a potential scenario where the GUI says one thing and the API says another if we later allow authors to pin their own brews.

File: server/homebrew.api.js [hunk 2]
```diff
@@ -188,8 +190,7 @@ const api = {
 			stub.renderer = stub.renderer || undefined; // Clear empty strings
 			stub = _.defaults(stub, DEFAULT_BREW_LOAD); // Fill in blank fields
 
-
-
+			stub.pinned = (stub?.pinned.length > 0);
```
Comment (calculuschild) [inline server/homebrew.api.js]: This seems like it would replace the pinned array with a boolean true/false at the time of loading. Is this intended?

File: client/homebrew/navbar/metadata.navitem.jsx [hunk 3]
```diff
@@ -27,6 +29,18 @@ const MetadataNav = createReactClass({
...
+		if(global.account) {
+			// Don't toggle if Brew Author
+			if(brew.authors.indexOf(global.account.username) > -1) return;
+			await this.setState((prevProps)=>({
+				pinnedBrew : !prevProps.pinnedBrew
+			}));
+			await request.put('/pin')
+				.send({ username: global.account.username,  shareId: this.props.brew.shareId, pin: this.state.pinnedBrew });
+		}
+	},
```
Comment (calculuschild) [inline client/homebrew/navbar/metadata.navitem.jsx:43]: One trick here, is `await` does not do anything with `this.setState`, since it doesn't return a promise. I believe you are trying to force the state to update before doing the server request, but React doesn't apply state changes until the next render pass anyway.

The more stable way to update the state and use that updated value inside the same function is to just calculate the new value and use that in all the needed places within this function, something like:

```
	togglePinState : async function(){
		if(global.account) {
			// Don't toggle if Brew Author
			if(brew.authors.indexOf(global.account.username) > -1) return;
			const newPinnedBrew = !this.state.pinnedBrew;
			      
			this.setState({pinnedBrew : newPinnedBrew});
			await request.put('/pin')
				.send({ username: global.account.username,  shareId: this.props.brew.shareId, pin: newPinnedBrew });
		}
	},
```

File: client/homebrew/navbar/metadata.navitem.jsx [hunk 4]
```diff
@@ -68,11 +82,17 @@ const MetadataNav = createReactClass({
 	},
 
 	render : function(){
-		return <Nav.item icon='fas fa-info-circle' color='grey' className='metadata'
-			onClick={()=>this.toggleMetaWindow()}>
-			{this.props.children}
-			{this.renderMetaWindow()}
-		</Nav.item>;
+		return <>
+			<Nav.item icon='fas fa-thumbtack' color='grey' className={`brewPin ${this.state.pinnedBrew ? 'active' : 'inactive'}`}
```
Comment (calculuschild) [inline client/homebrew/navbar/metadata.navitem.jsx]: This is OK for now, but it does seem like this pin button should be its own navitem.jsx component instead of nested inside of the metadata dropdown navitem.

File: server/homebrew.api.js [hunk 5]
```diff
@@ -240,7 +241,7 @@ const api = {
 	excludePropsFromUpdate : (brew)=>{
 		// Remove undesired properties
 		const modified = _.clone(brew);
-		const propsToExclude = ['_id', 'views', 'lastViewed'];
+		const propsToExclude = ['_id', 'views', 'lastViewed', 'pinned'];
```
Comment (calculuschild) [inline server/homebrew.api.js]: This is I assume to avoid overwriting someone's pins if they happen to pin while you are editing the document?

File: server/app.js [hunk 6]
```diff
@@ -460,6 +470,12 @@ export default async function createApp(vite) {
 			}
 		};
 
+		if( global.account) {
+			if(brew.pinned.indexOf(global.account.username) !== -1) brew.pinned = true;
```
Comment (calculuschild) [inline server/app.js]: I think this check will not work because in `homebrew.api.js getBrew()` we always force brew.pinned to be a boolean. We have already lost the list of users at this point.

File: server/homebrew.api.js [hunk 7]
```diff
@@ -104,7 +104,9 @@ const api = {
 
 		const userThemes = {};
 
-		const brews = await HomebrewModel.getByUser(username, true, fields, { tags: { $in: ['meta:theme', 'meta:Theme'] } });
+		const userBrewThemes = await HomebrewModel.getByUser(username, true, fields, { tags: { $in: ['meta:theme', 'meta:Theme'] }});
+		const pinnedBrewThemes = await HomebrewModel.getByUser(username, true, fields, { tags: { $in: ['meta:theme', 'meta:Theme'] }, pinned: { $in: [username] } });
```
Comment (calculuschild) [inline server/homebrew.api.js]: `getByUser` only finds brews where the current user is an author. It won't find pinned brews from other authors.

May need to make a new mongoose call `getByPinned` or something that is nearly identical but uses the pinned field instead of authors field.

**Review by calculuschild — CHANGES_REQUESTED** [review]
I approve of the implementation here. I am not able to test yet but I just noticed a few things that might cause bugs and had some clarifying questions.

I tried to keep the comments to *just* potential bugs. Someone can come along later and nitpick the design after this is merged.

File: server/homebrew.model.js (hunk 1 above)
Comment (dbolack-ab) [reply server/homebrew.model.js:102]: My thinking was that since this is an endpoint someone could hit it directly. 

If this is considered a minimal or non-concern, sure.

File: server/homebrew.api.js (hunk 2 above)
Comment (dbolack-ab) [reply server/homebrew.api.js]: Yes. 

I didn't think we wanted/needed to pass the array contents back to the sharepage and we don't need to keep evaluating it over and over.

If it makes things cleaner/clearer we could set the boolean in a different object member and null/drop the array from the loaded brew.

File: client/homebrew/navbar/metadata.navitem.jsx (hunk 3 above)
Comment (dbolack-ab) [reply client/homebrew/navbar/metadata.navitem.jsx:43]: I wondered about that. Makes sense.

File: client/homebrew/navbar/metadata.navitem.jsx (hunk 4 above)
Comment (dbolack-ab) [reply client/homebrew/navbar/metadata.navitem.jsx]: I tried it both ways, but due to the <metadata /> functionally being the title bar, I placed it there.

File: server/homebrew.api.js (hunk 5 above)
Comment (dbolack-ab) [reply server/homebrew.api.js]: I *believe* this will prevent the Author(s) from affecting the pinned list. That's the intent anyway.

File: server/homebrew.api.js (hunk 7 above)
Comment (dbolack-ab) [reply server/homebrew.api.js]: Oh, I forgot a small bit there to permit that with a null username.

A new Mongoose call works. i'll do that for clarity.

File: server/homebrew.api.js (hunk 2 above)
Comment (dbolack-ab) [reply server/homebrew.api.js]: Moved this to a new member and 'delete' the array when the share is loaded.

The array has been renamed to ***pinnedByUsers** and the new member is ***isPinned***

File: server/app.js (hunk 6 above)
Comment (dbolack-ab) [reply server/app.js]: That shouldn't happen in getBrew(). Cleaned up.

File: server/homebrew.api.js (hunk 7 above)
Comment (dbolack-ab) [reply server/homebrew.api.js]: Done

File: server/homebrew.api.js (hunk 2 above)
Comment (calculuschild) [reply server/homebrew.api.js]: Works for me

File: client/homebrew/navbar/navbar.less [hunk 8]
```diff
@@ -118,7 +118,14 @@
 			white-space      : nowrap;
 			background-color : transparent;		
 		}
-
+		&.brewPin {
+				&.active i {
+					color: red;
```
Comment (5e-Cleric) [inline client/homebrew/navbar/navbar.less:123]: Haven't seen the UI, but i recommend our @blue color from our colors.less in the shared folder for active icons.

File: server/app.js [hunk 9]
```diff
@@ -460,6 +481,15 @@ export default async function createApp(vite) {
...
 
+		if( global.account) {
+			if(brew.pinnedByUsers.indexOf(global.account.username) !== -1) brew.isPinned = true;
+			else { brew.isPinned = false; }
+		} else {
+			brew.isPinned = false;
+		}
+
+		delete brew.pinnedByUsers;
+
```
Comment (5e-Cleric) [inline server/app.js:492]: I might be wrong but won't this delete someone else's pinned brew, since all users will be stored within the one field?

File: server/brewDefaults.js [hunk 10]
```diff
@@ -2,32 +2,33 @@ import _ from 'lodash';
...
+	gDrive        : false,
+	trashed       : false,
+	bleedSize     : { top: '.125in', bottom: '.125in', inner: '.125in', outer: '.125in' },
+	safetySpace   : { top: '.25in', bottom: '.25in', outer: '.25in', inner: '.5in' },
+	trimSize      : { width: '8.5in', height: '11in' },
+	columns       : '2',
+	columnGutter  : '.125in',
+	license       : 'None',
+	legalAuthors  : '',
+	pinnedByUsers : []
```
Comment (5e-Cleric) [inline server/brewDefaults.js:31]: I think we could call this `pinnedBy`, no? not sure if we need it this long.

File: server/homebrew.model.js (hunk 1 above)
Comment (5e-Cleric) [reply server/homebrew.model.js:102]: I'd normally say it is a minimal concern, but i have been asked to add similar checks on every one of my PRs using an endpoint, so i have mixed feelings here. I say we keep it.

**Review by 5e-Cleric — COMMENTED** [review]
Overall looks good, left a few non-vital comments, will approve once i have properly tested.
